### PR TITLE
Move filtered SMJ right join out of `join_partial` phase

### PR DESCRIPTION
**Pull request by comphead**
## Which issue does this PR close?

Related #12359 .
Follow up on #12764 

## Rationale for this change
Same as #12764 but for Right Outer join

### CODE REVIEW

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 1]
```diff
@@ -746,21 +750,21 @@ fn last_index_for_row(
 // `false` - the row sent as NULL joined row
 fn get_corrected_filter_mask(
     join_type: JoinType,
-    indices: &UInt64Array,
-    ids: &[usize],
+    row_indices: &UInt64Array,
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:753]: renamed variables for readability

File: datafusion/sqllogictest/test_files/sort_merge_join.slt [hunk 2]
```diff
@@ -100,14 +100,13 @@ Alice 100 Alice 2
 Alice 50 Alice 1
 Alice 50 Alice 2
 
-# Uncomment when filtered RIGHT moved
 # right join with join filter
```
Comment (comphead) [inline datafusion/sqllogictest/test_files/sort_merge_join.slt:103]: reenabling tests

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 3]
```diff
@@ -149,16 +149,14 @@ async fn test_right_join_1k() {
...
-// https://github.com/apache/datafusion/issues/12359
 async fn test_right_join_1k_filtered() {
     JoinFuzzTestCase::new(
         make_staggered_batches(1000),
         make_staggered_batches(1000),
         JoinType::Right,
         Some(Box::new(col_lt_col_filter)),
     )
-    .run_test(&[JoinTestType::NljHj], false)
+    .run_test(&[JoinTestType::HjSmj, JoinTestType::NljHj], false)
```
Comment (alamb) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:159]: 🎉

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 4]
```diff
@@ -3643,102 +3664,12 @@ mod tests {
 
     #[tokio::test]
     async fn test_left_semi_join_filtered_mask() -> Result<()> {
-        let schema = Arc::new(Schema::new(vec![
```
Comment (alamb) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:3646]: why are these tests removed?

**Review by alamb — COMMENTED** [review]
Thanks @comphead  -- the only question I have is why some of the tests are removed

File: datafusion/physical-plan/src/joins/sort_merge_join.rs (hunk 4 above)
Comment (comphead) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:3646]: they are not removed, the common test data preparation was factored out in helper method

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 5]
```diff
@@ -1512,7 +1517,10 @@ impl SMJStream {
                     };
 
                     // Push the filtered batch which contains rows passing join filter to the output
-                    if matches!(self.join_type, JoinType::Left | JoinType::LeftSemi) {
+                    if matches!(
+                        self.join_type,
+                        JoinType::Left | JoinType::LeftSemi | JoinType::Right
```
Comment (alamb) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1522]: Do we also have to handle https://docs.rs/datafusion/latest/datafusion/common/enum.JoinType.html#variant.RightSemi 🤔

**Review by alamb — APPROVED** [review]
Thank you @comphead

File: datafusion/physical-plan/src/joins/sort_merge_join.rs (hunk 5 above)
Comment (comphead) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:1522]: At the end of the day all types of filtered joins will be moved.
I'm planning to cover this week RightSemi, LeftAnti, RightAnti and Full, lets see how it goes

**Pull request merged**